### ISSUE TITLE
Ability to Scale CustomResource 

### ISSUE THREAD

**Issue by michalxo** [fabric8io/kubernetes-client#4975]
### Is your enhancement related to a problem? Please describe

I am looking for an alternative of scaling custom resource like with following command:
```
kubectl scale activemqartemis/artemis-broker -n activemq-artemis-operator --replicas=0 # --replicas=10
```

Currently I see that only standard kubernetes resources are supported like deployment, statefulset, replicaset etc.
The thing is, that if I scaledown SS or RS our "deployment" uses, after `oc scale ss` operation finishes, out ClusterOperator reverts it back to previous state. (Which expects to call oc scale CR to update it's real value).

### Describe the solution you'd like

Ability to call something like 
`myGeneratedCRClient().inNamespace(namespace).resource(myResource).scale(int);`

**Comment by shawkins** [comment]
@michalxo yes, the current approach is that Resource<T> exposes only the minimal common functionality.  For custom resources that is limiting.  

A workaround is to directly call what is currently under the covers for the scale operation:  https://github.com/fabric8io/kubernetes-client/blob/05ff90ce67620b46397d238fa48b0df85987cea2/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/RollableScalableResourceOperation.java#L63

That will patch the number of replicas regardless of what the resource type is.

Another approach would be to make a DSL extension for your custom type - however that is not well supported at this point because the scale methods are not account for via the extension api.  You'd have to instead depend upon internal classes in the kuberentes-client jar.

Longer term the con of exposing additional operations directly on Resource is that it may dilute the value of the DSL - that is if you use the fully typed methods client.api().v1().kind().resource(value) - the current expectation is that only the relevant methods show up. 

In rough terms we can:
1. Make an exception for scaling operations and add that to Resource.
2. Go much further and in general make things more like kubectl which is operation, not resource, centric.  Which would greatly reduce the need / value of the typed DSL.
3. A compromise that would abstract the implementation and interfaces such that you can adapt a Resource to Scalable, Rollable, Loggable, etc.  This would make it more convenient to write generic logic, not just for custom types, but for operations across different versions client.resource(deploymentV1).adapt(Scaleable.class).scale(...) or client.resource(deploymentV1beta1).adapt(Scaleable.class).scale(...)

cc @manusa @rohanKanojia @csviri @metacosm

**Comment by metacosm** [comment]
I'm not sure I understand this issue. If I understand correctly, you have a CR (primary resource) that, once deployed to the cluster, triggers the creation of a scalable secondary resource (SS, RS, …). You would like to scale that secondary resource but when you do, the controller associated with your CR reverts its state to what it originally one. 

If what I described above is correct, then that is the expected behavior: a controller should control how the dependent resources associated with the primary resource behave. Presumably, the controller's implementer knows what should happen and the CR is the API to your service. It's the expected behavior for a controller to adjust the state of secondary resources that have deviated from the desired state (in your case because you interacted with them directly). 

If the number of replicas that are created for a secondary resource as part of you deploying your CR is not part of the CR "interface" then either that's a choice (meaning that the controller developer believe you shouldn't adjust them independently) or a bug (in which case, you should see with the operator's developer) but it's definitely not the role of the client to substitute to what controllers enforce.

**Comment by metacosm** [comment]
Ok, I completely misunderstood the issue 😭 but I think I do, now 🎉

**Comment by metacosm** [comment]
Perhaps we should have different traits that could be added to custom resources depending on which subresources they expose, which I think is what @shawkins's solution 3 is about. Being able to know when a resource has a status would definitely be helpful as well (there's an issue on this already, iirc).

**Comment by michalxo** [comment]
> A workaround is to directly call what is currently under the covers for the scale operation:
> 
> https://github.com/fabric8io/kubernetes-client/blob/05ff90ce67620b46397d238fa48b0df85987cea2/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/apps/v1/RollableScalableResourceOperation.java#L63
> 
> That will patch the number of replicas regardless of what the resource type is.
> 
Maybe I am doing something wrong,  but following command does nothing
```java
ResourceManager.getArtemisClient().resource(broker).patch(PatchContext.of(PatchType.JSON),
                "[{\"op\": \"replace\", \"path\":\"/spec/replicas\", \"value\":" + 4 + "}]");

```
This seems to do nothing for me. (No update on CRs, SS etc..)
Currently I can't provide any meaningful output, as my kubernetes is inaccessible for last few hours now (up/down connectivity issues). :)

**Comment by shawkins** [comment]
@michalxo sorry, the existing patch is against the spec, not the scale sub-resource.  The mechanism to easily target a sub-resource is not exposed via the public api currently.

**Comment by michalxo** [comment]
Thank you! Looking forward to 6.6.0 release :)
